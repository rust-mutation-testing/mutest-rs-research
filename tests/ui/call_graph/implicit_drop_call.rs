//@ print-call-graph
//@ print-targets
//@ stdout
//@ stderr: empty

struct ImplsDrop;

fn drop_impl() {}
impl Drop for ImplsDrop {
    fn drop(&mut self) {
        drop_impl();
    }
}

#[test]
fn test_implicit_drop_call() {
    let _ = ImplsDrop;
}

fn drop_impls_drop() {
    let _ = ImplsDrop;
}

#[test]
fn test_implicit_indirect_drop_call() {
    drop_impls_drop();
}

// NOTE: Same as `core::mem::drop`, but we explicitly define it for the sake of the test.
fn implicit_drop<T>(_: T) {}

#[test]
fn test_implicit_generic_drop_call() {
    implicit_drop(ImplsDrop);
}

struct NoDrop;

#[test]
fn test_no_implicit_drop_call() {
    let _ = NoDrop;
}
