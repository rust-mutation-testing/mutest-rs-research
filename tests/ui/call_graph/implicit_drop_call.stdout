
@@@ call graph @@@

entry points:

test test_implicit_drop_call
  -> <ImplsDrop as std::ops::Drop>::drop at tests/ui/call_graph/implicit_drop_call.rs:10:5: 10:23 (#0)
test test_implicit_generic_drop_call
  -> implicit_drop::<ImplsDrop> at tests/ui/call_graph/implicit_drop_call.rs:30:1: 30:26 (#0)
test test_implicit_indirect_drop_call
  -> drop_impls_drop at tests/ui/call_graph/implicit_drop_call.rs:20:1: 20:21 (#0)
test test_no_implicit_drop_call

nested calls at distance 1:

<ImplsDrop as std::ops::Drop>::drop at tests/ui/call_graph/implicit_drop_call.rs:10:5: 10:23 (#0)
  -> drop_impl at tests/ui/call_graph/implicit_drop_call.rs:8:1: 8:15 (#0)
drop_impls_drop at tests/ui/call_graph/implicit_drop_call.rs:20:1: 20:21 (#0)
  -> <ImplsDrop as std::ops::Drop>::drop at tests/ui/call_graph/implicit_drop_call.rs:10:5: 10:23 (#0)
implicit_drop::<ImplsDrop> at tests/ui/call_graph/implicit_drop_call.rs:30:1: 30:26 (#0)
  -> <ImplsDrop as std::ops::Drop>::drop at tests/ui/call_graph/implicit_drop_call.rs:10:5: 10:23 (#0)

@@@ targets @@@

tests -(1)-> drop_impl at tests/ui/call_graph/implicit_drop_call.rs:8:1: 8:15 (#0)
  (1) test_implicit_drop_call
  (2) test_implicit_generic_drop_call
  (2) test_implicit_indirect_drop_call

tests -(0)-> <ImplsDrop as std::ops::Drop>::drop at tests/ui/call_graph/implicit_drop_call.rs:10:5: 10:23 (#0)
  (0) test_implicit_drop_call
  (1) test_implicit_generic_drop_call
  (1) test_implicit_indirect_drop_call

tests -(0)-> drop_impls_drop at tests/ui/call_graph/implicit_drop_call.rs:20:1: 20:21 (#0)
  (0) test_implicit_indirect_drop_call

tests -(0)-> implicit_drop at tests/ui/call_graph/implicit_drop_call.rs:30:1: 30:26 (#0)
  (0) test_implicit_generic_drop_call

targets: 4 total; 4 safe; 0 unsafe (0 tainted)
